keps/sig-testing/1746-stage-e2e-test-framework/README.md

```diff
@@ -0,0 +1,643 @@
+of what we already have instead of starting from scratch.
+
+Moving the e2e test framework into staging means that the e2e test framework
+will be used and maintained as a component independent of core Kubernetes.
+By doing this, the e2e test framework will only make use of public APIs and
+will be self-contained.
+It will also enable out-of-tree projects to import the e2e test framework
+without importing all of Kubernetes.
+
+Moving the e2e test framework to staging will also help maintainers of e2e tests
+have tools to easily test any component that builds on Kubernetes in the same
+manner as core Kubernetes.
+People will be able to use all the same utility functions used in e2e tests in
+core Kubernetes and will be able to use the e2e test framework's context to configure
+their e2e test environments (i.e., manage information about the cluster being
+used for e2e tests, the provider, ETCD).
+
+This will also help external consumers of the e2e test framework by reducing
+the amount of dependency management they have to do now when importing the e2e
+test framework.
+
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The e2e test framework is a Kubernetes component that has been evolving
+organically to meet all the needs that contributors have in order to test
+enhancements and behaviours of Kubernetes.
+The SIG Testing sub-project
+[Testing Commons](https://github.com/kubernetes/community/tree/master/sig-testing#testing-commons)
+has been looking at ways of improving the user experience for contributors to
+the Kubernetes community.
+This KEP is part of that goal.
+
+We foresee the following benefits coming out of this work:
+
+1. Maintainability: the e2e test framework lacks a general API - there has been
+  little discussion on what parts of the framework do what.
+  From this work we plan to organize the code that is present and use only
+  external libraries and APIs available for e2e tests.
+  This decoupling will ensure the e2e test framework packages are better
+  contained allowing for e2e tests to have a more standard flow.
+2. External consumption: The Kubernetes ecosystem has grown considerably since
+  the creation of the e2e test framework. Some projects, such as the cloud
+  providers and CSI extensions, began within the core Kubernetes codebase and
+  have been migrated out or are in the process of being migrated out. We want
+  to provide them with the same set level of functionality they had in core
+  without having to deal with the problems of importing the entirety of core
+  Kubernetes.
```

Comment:
Quick question: Can we expect actual external consumers of e2e framework soon after initial staging release?
To improve the framework, it is nice to get feedback from the consumers quickly.